### ISSUE TITLE
one HTTP forbidden error causes program to abort

### ISSUE THREAD

**Issue by doctorcolossus** [thetric/ilias-downloader-cli#11]
I don't know why, but ilias-downloader-cli can never seem to access this particular file:

```
Downloading file "Lösung-Bonusklausur2-MedStat-WS12-13.pdf" (16,500 bytes)... 
Exception in thread "main" com.github.thetric.iliasdownloader.service.exception.IliasException: Failed to GET https://www.ilias.fh-dortmund.de/ilias/webdav.php/ilias-fhdo/ref_9538//Alte_Klausuren//Miniklausuren//L%C3%B6sung-Bonusklausur2-MedStat-WS12-13.pdf: Forbidden
        at com.github.thetric.iliasdownloader.service.webparser.impl.webclient.OkHttpIliasWebClient.checkResponse(OkHttpIliasWebClient.kt:90)
        at com.github.thetric.iliasdownloader.service.webparser.impl.webclient.OkHttpIliasWebClient.getAsInputStream(OkHttpIliasWebClient.kt:84)
        at com.github.thetric.iliasdownloader.service.webparser.impl.WebIliasService.getContentAsStream(WebIliasService.kt:22)
        at com.github.thetric.iliasdownloader.cli.sync.ItemDownloadingItemVisitor.syncAndSaveFile(ItemDownloadingItemVisitor.kt:82)
        at com.github.thetric.iliasdownloader.cli.sync.ItemDownloadingItemVisitor.handleFile(ItemDownloadingItemVisitor.kt:48)
        at com.github.thetric.iliasdownloader.service.webparser.impl.course.CourseSyncServiceImpl.walkIliasItemNode(CourseSyncServiceImpl.kt:107)
        at com.github.thetric.iliasdownloader.service.webparser.impl.course.CourseSyncServiceImpl.visit(CourseSyncServiceImpl.kt:50)
        at com.github.thetric.iliasdownloader.service.webparser.impl.course.CourseSyncServiceImpl.walkIliasItemNode(CourseSyncServiceImpl.kt:103)
        at com.github.thetric.iliasdownloader.service.webparser.impl.course.CourseSyncServiceImpl.visit(CourseSyncServiceImpl.kt:50)
        at com.github.thetric.iliasdownloader.service.webparser.impl.course.CourseSyncServiceImpl.walkIliasItemNode(CourseSyncServiceImpl.kt:103)
        at com.github.thetric.iliasdownloader.service.webparser.impl.course.CourseSyncServiceImpl.visit(CourseSyncServiceImpl.kt:50)
        at com.github.thetric.iliasdownloader.service.webparser.impl.WebIliasService.visit(WebIliasService.kt:38)
        at com.github.thetric.iliasdownloader.cli.SyncController.executeSync(SyncController.kt:39)
        at com.github.thetric.iliasdownloader.cli.SyncController.startSync(SyncController.kt:28)
        at com.github.thetric.iliasdownloader.cli.CliController.startCliController(CliController.kt:41)
        at com.github.thetric.iliasdownloader.cli.CliKt.main(Cli.kt:30)
```

That wouldn't be so bad, especially since I encounter no trouble attempting to download this file manually through my browser. However, because it causes ilias-downloader-cli to abort, I am unable to finish syncing the other files in this course and all the remaining ones.

After downloading the file manually, I tried manually changing the timestamp of the file to trick ilias-downloader-cli into thinking it had already been synced, but that didn't work either. Actually, I noticed that all the timestamps of local synced files seem to be off +/- 5 minutes from the timestamps displayed in the Ilias web interface, but the difference appears to be somewhat random, so I can't guess the exact time ilias-downloader-cli expects the file to have and I am not sure whether it is comparing only minutes, or also seconds or even milliseconds, so I have given up on that workaround for now. I suppose I could dig into the sourcecode, but I am not that motivated and am hoping you guys can offer me an easier solution or workaround.

I would like if the program caught HTTP errors and prompted the user to abort, retry, or skip the current file. It would also be nice if there were an ignored files list in the configuration json. It would also be nice to figure out what causes the HTTP forbidden error, since this doesn't happen for me in my browser, so let me know if you have any ideas about how we could get to the bottom of that. Thanks.

**Comment by thetric** [comment]
This is indeed a very annoying issue.

Could you please check if you can access https://www.ilias.fh-dortmund.de/ilias/webdav.php/ilias-fhdo/ref_9538//Alte_Klausuren//Miniklausuren//L%C3%B6sung-Bonusklausur2-MedStat-WS12-13.pdf in your browser? Perhaps you need to login first.

Are you using the latest version (3.0.0)?

**Comment by doctorcolossus** [comment]
> Are you using the latest version (3.0.0)?
Yes, I am using the 3.0.0 release downloaded from [here](https://github.com/thetric/ilias-downloader-cli/releases).

When I attempt to open that link in my browser (Chromium), after logging in, I also receive the same error:
```
Access to www.ilias.fh-dortmund.de was denied
You don't have authorization to view this page.
HTTP ERROR 403
```

Navigating to this file manually in my browser, the download link appears to be `https://www.ilias.fh-dortmund.de/ilias/goto_ilias-fhdo_file_474343_download.html`.

**Comment by thetric** [comment]
The CLI uses a different URL (not the same as the 'normal' web UI) to find the courses contents and to download the files. Can you check if you are able access the parent directories of the file?

* https://www.ilias.fh-dortmund.de/ilias/webdav.php/ilias-fhdo/ref_9538//Alte_Klausuren//Miniklausuren
* https://www.ilias.fh-dortmund.de/ilias/webdav.php/ilias-fhdo/ref_9538//Alte_Klausuren
* https://www.ilias.fh-dortmund.de/ilias/webdav.php/ilias-fhdo/ref_9538

I can't figure out why you get the HTTP 403 Forbidden error. My first thought was that the access was denied because you haven't joined the course, but this isn't true as you can download the file via the web interface. If some of the links above are working I might have a start for further investigations.
Otherwise I would start implementing your requested feature to exclude files.

> It would also be nice if there were an ignored files list in the configuration json

Ignoring these "broken" files wouldn't be a great solution in your case because you still would have to download the ignored files by yourself (that should be the task of the CLI :sweat_smile:)

**Comment by doctorcolossus** [comment]
I am able to access all three parent directories and all 33 of the other files in 'Miniklausuren'. However, clicking on 'Lösung-Bonusklausur2-MedStat-WS12-13.pdf' from that file listing in my browser does give me the same HTTP 403 unauthorized access error, so perhaps this is a bug in ILIAS. However, as I mentioned above, I am able to download the file in question through the 'normal' web UI.

Yes, I agree that the task of your CLI is to automate downloading files. However, I still think the program should somehow be able to handle unexpected errors like this and continue its progress. If there are 100 files and the 10th one gives such an unauthorized access error, wouldn't the user still be better off having the last 90 files instead of only the first 9? My workaround for now is to disable synchronization of this entire folder. That is the only way I can keep my other classes synchronized for now.

**Comment by thetric** [comment]
As a temporary "fix" for the `403 Forbidden` I released [3.0.1](https://github.com/thetric/ilias-downloader-cli/releases/tag/3.0.1) which continues on failed download attempts.

Just to be sure that it isn't my mistake could you please test [another Ilias downloader](http://www.whiledo.de/index.php?p=iliasdownloader) (by a colleague of mine)? If it also fails to download the file it would be a stronger evidence that it _is_ an Ilias issue.

**Comment by doctorcolossus** [comment]
Awesome, thank you!
However, with this new version I get a different error on the same file:
```
Downloading file "Lösung-Bonusklausur2-MedStat-WS12-13.pdf" (16,500 bytes)... 
Exception in thread "main" java.util.MissingResourceException: Can't find resource for bundle java.util.PropertyResourceBundle, key sync.download.course.failed
        at java.base/java.util.ResourceBundle.getObject(ResourceBundle.java:544)
        at java.base/java.util.ResourceBundle.getString(ResourceBundle.java:501)
        at com.github.thetric.iliasdownloader.cli.sync.ItemDownloadingItemVisitor.getLocalizedMessage(ItemDownloadingItemVisitor.kt:56)
        at com.github.thetric.iliasdownloader.cli.sync.ItemDownloadingItemVisitor.syncAndSaveFile(ItemDownloadingItemVisitor.kt:95)
        at com.github.thetric.iliasdownloader.cli.sync.ItemDownloadingItemVisitor.handleFile(ItemDownloadingItemVisitor.kt:49)
        at com.github.thetric.iliasdownloader.service.webparser.impl.course.CourseSyncServiceImpl.walkIliasItemNode(CourseSyncServiceImpl.kt:107)
        at com.github.thetric.iliasdownloader.service.webparser.impl.course.CourseSyncServiceImpl.visit(CourseSyncServiceImpl.kt:50)
        at com.github.thetric.iliasdownloader.service.webparser.impl.course.CourseSyncServiceImpl.walkIliasItemNode(CourseSyncServiceImpl.kt:103)
        at com.github.thetric.iliasdownloader.service.webparser.impl.course.CourseSyncServiceImpl.visit(CourseSyncServiceImpl.kt:50)
        at com.github.thetric.iliasdownloader.service.webparser.impl.course.CourseSyncServiceImpl.walkIliasItemNode(CourseSyncServiceImpl.kt:103)
        at com.github.thetric.iliasdownloader.service.webparser.impl.course.CourseSyncServiceImpl.visit(CourseSyncServiceImpl.kt:50)
        at com.github.thetric.iliasdownloader.service.webparser.impl.WebIliasService.visit(WebIliasService.kt:38)
        at com.github.thetric.iliasdownloader.cli.SyncController.executeSync(SyncController.kt:39)
        at com.github.thetric.iliasdownloader.cli.SyncController.startSync(SyncController.kt:28)
        at com.github.thetric.iliasdownloader.cli.CliController.startCliController(CliController.kt:41)
        at com.github.thetric.iliasdownloader.cli.CliKt.main(Cli.kt:30)
```

My first attempt to get your colleague's program was unsuccessful. When I try the import courses feature, I get an indescript "Fehler bei Erkennung" and on the command line output I see a cookie rejected warning:
```
Apr 17, 2018 10:32:25 PM org.apache.http.client.protocol.ResponseProcessCookies processCookies
WARNING: Cookie rejected: "[version: 0][name: ilClientId][value: ilias-fhdo][domain: il_cookie_domain][path: IL_COOKIE_PATH][expiry: null]". Illegal path attribute "IL_COOKIE_PATH". Path of origin: "/ilias/login.php"
```

**Comment by thetric** [comment]
Published [3.0.2](https://github.com/thetric/ilias-downloader-cli/releases/tag/3.0.2) which addresses the `MisingResourceException`.

The automatic course recognition of his program doesn't work. You have to add the course manually by following the instructions in the dialog under `Kurse > Kurse hinzufügen...`

**Comment by thetric** [comment]
Closing as the issue is resolved. The 403 Forbidden error seems to be an issue with the ilias server. The ilias bug tracker is password protected so I can neither search their bug reports nor create a new one :disappointed:
